actionview/lib/action_view/helpers/tag_helper.rb

```diff
@@ -26,7 +26,42 @@ module TagHelper
       PRE_CONTENT_STRINGS[:textarea]  = "\n"
       PRE_CONTENT_STRINGS["textarea"] = "\n"
 
+      class TagBuilder #:nodoc:
+        include TagHelper
 
+        VOID_ELEMENTS = %w(base br col embed hr img input keygen link meta
+        param source track wbr).to_set
+
+        VOID_ELEMENTS.merge(VOID_ELEMENTS.map(&:to_sym))
+
+        def initialize(view_context)
+          @view_context = view_context
+        end
+
+        private
+          def tag_string(name, content_or_options = nil, options_or_escape = nil, escape = true, &block)
+            case
+            when block_given?
+              content_tag_string(name, @view_context.capture(self, &block), content_or_options, options_or_escape)
+            when content_or_options.is_a?(String)
+              content_tag_string(name, content_or_options, options_or_escape, escape)
+            when VOID_ELEMENTS.include?(name)
+              options_or_escape = true if options_or_escape.nil?
+              "<#{name}#{tag_options(content_or_options, options_or_escape)}>".html_safe
+            else
+              options_or_escape = true if options_or_escape.nil?
+              content_tag_string(name, "", content_or_options, options_or_escape)
+            end
+          end
+
+          def method_missing(called, *args, &block)
+            tag_string(called, *args, &block)
+          end
+      end
+
+      # Returns an HTML tag. Supports two syntax variants: traditonal and modern.
```

Comment:
Let's flip this, so that the modern builder style comes first; that's what we want people to be using from now on. We should rename traditional to legacy to strongly discourage it.